File: bridges/atas/original/tlade_bridge_atas.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT
# Copyright (c) 2026 Mihai Ostafe — TLADe ATAS Bridge contribution
# Copyright (c) 2026 TLADe — Trade Like a Dealer (https://tradelikeadealer.com)
"""
TLADe Bridge — ATAS Edition  v1.2.0

Python receiver for the TLAdeBridgeATAS.cs indicator running inside ATAS.
Speaks the standard TLADe Bridge protocol on port 5000, exposing /health
and /ib_data to the TLADe Terminal.

Endpoints:
  POST /push_spot    <- ATAS indicator pushes every tick
  POST /push_bar     <- ATAS indicator pushes the closed 5m bar
  POST /push_daily   <- ATAS indicator pushes the daily bar
  POST /push_levels  <- external source pushes GEX levels (Call Wall, Put Wall, etc.)
  GET  /health       -> TLADe Terminal liveness probe
  GET  /ib_data      -> TLADe Terminal pulls live data (with real chart_data)
  GET  /get_levels   -> ATAS-side TLAdeLevels indicator pulls stored GEX levels

Run:
  pip install flask flask-cors
  python tlade_bridge_atas.py

/push_levels payload format:
  POST /push_levels
  {
    "ticker": "ES",
    "levels": [
      {"price": 5300.00, "type": "call_wall_strong", "label": "5300 CW Call Wall [S]"},
      {"price": 5250.00, "type": "call_wall",         "label": "5250 CW Call Wall"},
      {"price": 5200.00, "type": "put_wall",          "label": "5200 PW Put Wall"},
      {"price": 5180.00, "type": "max_pain",          "label": "5180 Max Pain"},
      {"price": 5270.00, "type": "em_high",           "label": "5270 EM High"}
    ]
  }

Accepted level types:
  call_wall_strong, call_wall, put_wall_strong, put_wall,
  max_pain, em_high, em_low, vwap_session, zg_level,
  vol_band_up, vol_band_down
"""

from datetime import datetime, timezone
from threading import Lock

from flask import Flask, jsonify, request
from flask_cors import CORS

VERSION = "1.2.0"
PORT    = 5000
HEARTBEAT_WINDOW_S = 15   # a tick is considered "fresh" if seen within the last N seconds
MAX_BARS           = 500  # how many 5m bars we keep in memory per ticker

app = Flask(__name__)
CORS(app, resources={r"/*": {"origins": "*"}})

_lock  = Lock()

# ── Per-ticker state ──────────────────────────────────────────────────────────
# {
#   "ES": {
#     "spot":  {"price": float, "ts": str, "received_at": datetime},
#     "bars":  [ {time, open, high, low, close, volume, ask_volume, bid_volume,
#                 delta, max_delta, min_delta}, ... ],   # max MAX_BARS
#     "daily": [ {time, open, high, low, close, volume, delta}, ... ]
#   }
# }
_state = {}

# Per-ticker GEX levels: {"ES": {"levels": [...], "ts": "..."}, ...}
_levels = {}

TICKERS = ["ES", "NQ"]

def _get_state(ticker: str) -> dict:
    if ticker not in _state:
        _state[ticker] = {"spot": None, "bars": [], "daily": []}
    return _state[ticker]


def _get_levels(ticker: str) -> dict:
    if ticker not in _levels:
        _levels[ticker] = {"levels": [], "ts": None}
    return _levels[ticker]


def _normalize(raw: str) -> str:
    s = (raw or "").upper()
    return "NQ" if ("NQ" in s or "NDX" in s or "MNQ" in s) else "ES"


def _is_fresh(received_at) -> bool:
    if received_at is None:
        return False
    return (datetime.now(timezone.utc) - received_at).total_seconds() < HEARTBEAT_WINDOW_S


# ─────────────────────────────────────────────────────────────────────────────
#  INBOUND — the ATAS indicator pushes data here
# ─────────────────────────────────────────────────────────────────────────────

@app.route("/push_spot", methods=["POST"])
def push_spot():
    """Live tick from the ATAS indicator (1:1 compatible with the NT8 bridge)."""
    data = request.get_json(silent=True)
    if not data or "ticker" not in data or "spot" not in data:
        return jsonify({"error": "invalid payload — expected {ticker, spot, ts}"}), 400
    try:
        price = float(data["spot"])
    except (TypeError, ValueError):
        return jsonify({"error": "spot must be numeric"}), 400

    ticker = _normalize(str(data["ticker"]))
    now    = datetime.now(timezone.utc)

    with _lock:
        st = _get_state(ticker)
        st["spot"] = {
            "price":       price,
            "ts":          data.get("ts") or now.isoformat(),
            "received_at": now,
            "provider":    data.get("provider", "unknown"),
        }

    print(f"[SPOT]  {ticker} -> {price}  ({data.get('provider','?')})")
    return jsonify({"ok": True, "ticker": ticker})


@app.route("/push_bar", methods=["POST"])
def push_bar():
    """Closed 5m bar with full order-flow fields."""
    data = request.get_json(silent=True)
    if not data or "ticker" not in data:
        return jsonify({"error": "invalid payload"}), 400

    ticker = _normalize(str(data["ticker"]))

    bar = {
        "time":       data.get("time"),
        "open":       data.get("open"),
        "high":       data.get("high"),
        "low":        data.get("low"),
        "close":      data.get("close"),
        "volume":     data.get("volume"),
        "ask_volume": data.get("ask_volume"),
        "bid_volume": data.get("bid_volume"),
        "delta":      data.get("delta"),
        "max_delta":  data.get("max_delta"),
        "min_delta":  data.get("min_delta"),
    }

    with _lock:
        st = _get_state(ticker)
        # avoid duplicates (same bar_index)
        idx = data.get("bar_index")
        if idx is not None and st["bars"] and st["bars"][-1].get("bar_index") == idx:
            st["bars"][-1] = {**bar, "bar_index": idx}
        else:
            st["bars"].append({**bar, "bar_index": idx})
            if len(st["bars"]) > MAX_BARS:
                st["bars"] = st["bars"][-MAX_BARS:]

    delta_str = f"D={bar['delta']:+.0f}" if bar.get("delta") is not None else ""
    print(f"[BAR]   {ticker} {bar.get('time','?')}  C={bar.get('close')}  Vol={bar.get('volume')}  {delta_str}")
    return jsonify({"ok": True, "ticker": ticker, "bars_stored": len(_state[ticker]["bars"])})


@app.route("/push_daily", methods=["POST"])
def push_daily():
    """Daily (D1) bar from ATAS."""
    data = request.get_json(silent=True)
    if not data or "ticker" not in data:
        return jsonify({"error": "invalid payload"}), 400

    ticker = _normalize(str(data["ticker"]))
    bar    = {k: data.get(k) for k in ("time","open","high","low","close","volume","ask_volume","bid_volume","delta")}

    with _lock:
        st = _get_state(ticker)
        # replace today's bar if already present
        if st["daily"] and st["daily"][-1].get("time","")[:10] == (bar.get("time") or "")[:10]:
            st["daily"][-1] = bar
        else:
            st["daily"].append(bar)
            if len(st["daily"]) > 365:
                st["daily"] = st["daily"][-365:]

    print(f"[DAILY] {ticker} {bar.get('time','?')[:10]}  C={bar.get('close')}  Vol={bar.get('volume')}")
    return jsonify({"ok": True, "ticker": ticker})


# ─────────────────────────────────────────────────────────────────────────────
#  OUTBOUND — the TLADe Terminal pulls data from here
# ─────────────────────────────────────────────────────────────────────────────

@app.route("/health", methods=["GET"])
def health():
    """Liveness probe (Bridge Protocol §health)."""
    with _lock:
        live = [k for k, v in _state.items() if v["spot"] and _is_fresh(v["spot"]["received_at"])]

    return jsonify({
        "status":       "ok",
        "tlade_bridge": True,
        "ib_connected": len(live) > 0,
        "provider":     "atas",
        "lite_mode":    False,       # we have real chart_data, not lite
        "version":      VERSION,
        "live_tickers": live,
        "port":         PORT,
    })


@app.route("/ib_data", methods=["GET"])
def ib_data():
    """
    Full payload for the terminal (Bridge Protocol §ib_data).
    Unlike the NT8 bridge, we return a real chart_data block with historical
    5m bars + delta, which allows TLADe to render the chart from ATAS data.
    """
    ticker = _normalize(request.args.get("ticker", "ES"))

    with _lock:
        st = _get_state(ticker)
        spot = st.get("spot")
        fresh = bool(spot and _is_fresh(spot["received_at"]))

        if not fresh:
            return jsonify({"error": f"no recent tick for {ticker}"}), 404

        bars   = list(st["bars"])
        daily  = list(st["daily"])
        price  = spot["price"]
        ts     = spot["ts"]

    # ── Serialize 5m bars into chart_data format ───────────────────────────
    def extract(key, lst):
        return [b.get(key) for b in lst]

    chart_data = {
        "time":       extract("time",   bars),
        "open":       extract("open",   bars),
        "high":       extract("high",   bars),
        "low":        extract("low",    bars),
        "close":      extract("close",  bars),
        "volume":     extract("volume", bars),
        # extra fields vs the NT8 bridge:
        "ask_volume": extract("ask_volume", bars),
        "bid_volume": extract("bid_volume", bars),
        "delta":      extract("delta",      bars),
        "max_delta":  extract("max_delta",  bars),
        "min_delta":  extract("min_delta",  bars),
    }

    daily_data = {
        "time":   extract("time",   daily),
        "open":   extract("open",   daily),
        "high":   extract("high",   daily),
        "low":    extract("low",    daily),
        "close":  extract("close",  daily),
        "volume": extract("volume", daily),
        "delta":  extract("delta",  daily),
    }

    return jsonify({
        "chart_data":  chart_data,   # 5m bars with order flow
        "daily_data":  daily_data,   # D1 bars
        "live_price":  price,
        "last_price":  price,
        "instrument":  ticker,
        "ts":          ts,
        "provider":    "atas",
        "bars_count":  len(bars),
        "daily_count": len(daily),
    })


# ─────────────────────────────────────────────────────────────────────────────
#  GEX LEVELS — pushed by an external source / pulled by the ATAS indicator
# ─────────────────────────────────────────────────────────────────────────────

@app.route("/push_levels", methods=["POST"])
def push_levels():
    """
    Receive GEX levels from an external source (Python script, browser script,
    TLADe Terminal wrapper, etc.) and store them for the ATAS indicator.

    JSON body:
      {
        "ticker": "ES",
        "levels": [
          {"price": 5300.0, "type": "call_wall_strong", "label": "5300 CW Call Wall [S]"},
          {"price": 5250.0, "type": "call_wall",         "label": "5250 CW Call Wall"},
          ...
        ]
      }
    """
    data = request.get_json(silent=True)
    if not data or "levels" not in data:
        return jsonify({"error": "invalid payload — expected {ticker, levels:[{price,type,label}]}"}), 400

    raw_levels = data.get("levels", [])
    if not isinstance(raw_levels, list):
        return jsonify({"error": "levels must be an array"}), 400

    ticker = _normalize(str(data.get("ticker", "ES")))
    now    = datetime.now(timezone.utc)

    # Validate + normalize
    validated = []
    for item in raw_levels:
        try:
            price = float(item.get("price", 0))
            if price == 0:
                continue
            validated.append({
                "price": round(price, 4),
                "type":  str(item.get("type", "unknown")),
                "label": str(item.get("label", f"{price:.2f}")),
            })
        except (TypeError, ValueError):
            continue

    # Sort descending by price (handy for display)
    validated.sort(key=lambda x: x["price"], reverse=True)

    with _lock:
        lv = _get_levels(ticker)
        lv["levels"] = validated
        lv["ts"]     = now.isoformat()

    print(f"[LEVELS] {ticker} -> {len(validated)} levels received")
    return jsonify({"ok": True, "ticker": ticker, "count": len(validated)})


@app.route("/get_levels", methods=["GET"])
def get_levels():
    """
    Return the stored GEX levels for a ticker.
    Used by the ATAS-side TLAdeLevels.cs indicator.

    Query params:
      ticker=ES|NQ  (default: ES)

    Response:
      {
        "ticker":  "ES",
        "count":   15,
        "ts":      "2026-05-01T20:00:00Z",
        "levels":  [ {"price": 5300.0, "type": "call_wall_strong", "label": "5300 CW Call Wall [S]"}, ... ]
      }
    """
    ticker = _normalize(request.args.get("ticker", "ES"))
    with _lock:
        lv = _get_levels(ticker)
        return jsonify({
            "ticker":  ticker,
            "count":   len(lv["levels"]),
            "ts":      lv["ts"],
            "levels":  list(lv["levels"]),
        })


# ─────────────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    print(f"TLADe Bridge - ATAS Edition  v{VERSION}")
    print(f"Listening on http://localhost:{PORT}  (auto-detected by TLADe Terminal)")
    print()
    print("  ATAS -> bridge:")
    print(f"    POST /push_spot   (live tick)")
    print(f"    POST /push_bar    (5m bar)")
    print(f"    POST /push_daily  (daily bar)")
    print(f"    POST /push_levels (GEX levels)")
    print()
    print("  Bridge -> TLADe Terminal:")
    print(f"    GET  /health")
    print(f"    GET  /ib_data?ticker=ES")
    print()
    print("  Bridge -> ATAS TLAdeLevels indicator:")
    print(f"    GET  /get_levels?ticker=ES")
    print()
    app.run(host="0.0.0.0", port=PORT, debug=False, threaded=True)

File: bridges/atas/original/TLAdeBridgeATAS.cs
// ============================================================
//  SPDX-License-Identifier: MIT
//  Copyright (c) 2026 Mihai Ostafe — TLADe ATAS Bridge contribution
//  Copyright (c) 2026 TLADe — Trade Like a Dealer (https://tradelikeadealer.com)
//
//  TLADe Bridge — ATAS Edition  v2.4.0
//  C# indicator for ATAS (Advanced Time & Sales)
//
//  Architecture:
//    Rithmic / CQG feed
//      -> ATAS
//           -> TLAdeBridgeATAS.cs   (indicator, HTTP POST push)
//                -> tlade_bridge_atas.py   (Python receiver, port 5000)
//                     -> TLADe Terminal    (auto-detects localhost:5000)
//
//  Push endpoints (ATAS -> receiver):
//    POST /push_spot    -> live tick
//    POST /push_bar     -> closed bar (OHLCV + delta + bid/ask volume)
//    POST /push_daily   -> daily bar (last bar of the previous trading day)
//
//  Install:
//    1. Build:   dotnet build TLAdeBridgeATAS.csproj -c Release
//                (the DLL is auto-copied to Documents\ATAS\Scripts\)
//    2. ATAS:    Settings -> Extensions -> Reload
//    3. Add the indicator to an ES or NQ chart
//    4. Start the receiver: double-click start.bat (or `python tlade_bridge_atas.py`)
//    5. TLADe Terminal auto-detects the bridge on localhost:5000.
//
//  v2.4.0: Channel<> for bars — single worker thread, no thread-pool explosion
//    - _postedBarTimes: every bar sent exactly once (key = candle.Time.Ticks)
//    - _barSem: at most 3 concurrent POSTs to Flask
//    - removed the broken time/index filter
// ============================================================

using System;
using System.Collections.Concurrent;
using System.IO;
using System.Net.Http;
using System.Text;
using System.Threading;
using System.Threading.Channels;
using System.Threading.Tasks;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using ATAS.Indicators;

namespace ATAS.Indicators.Technical
{
    [DisplayName("TLADe Bridge")]
    [Description("Trimite date live (ticks + bare + delta) catre TLADe Terminal pe localhost:5000")]
    public class TLAdeBridgeATAS : Indicator
    {
        private static readonly HttpClient _http = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(5)
        };

        private static readonly string _logPath = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments),
            "ATAS", "tlade_bridge.log");

        // Channel with capacity 600: a single worker sends bars sequentially
        // Previne thread pool explosion la flood de bare istorice
        private readonly Channel<(string ticker, int bar, CandleData cd)> _barChannel
            = Channel.CreateBounded<(string, int, CandleData)>(new BoundedChannelOptions(600)
              {
                  FullMode    = BoundedChannelFullMode.DropOldest,
                  SingleReader = true,
                  SingleWriter = false
              });

        // Bar deduplication: every bar sent exactly once (key = candle.Time.Ticks)
        private readonly ConcurrentDictionary<long, byte> _postedBarTimes
            = new ConcurrentDictionary<long, byte>();

        // State intern
        private double   _lastPostedPrice    = -1;
        private DateTime _lastSpotTime       = DateTime.MinValue;
        private int      _lastPostedDailyBar = -1;
        private DateTime _lastBarDate        = DateTime.MinValue.Date;
        private bool     _liveMode           = false;

        [Display(Name = "Port receiver", GroupName = "TLADe", Order = 1)]
        [Range(1024, 65535)]
        public int Port { get; set; } = 5000;

        [Display(Name = "Interval minim intre tickuri (sec)", GroupName = "TLADe", Order = 2)]
        [Range(1, 30)]
        public int SpotThrottleSeconds { get; set; } = 2;

        [Display(Name = "Trimite bare intraday (off = NT8-like, doar SPOT)", GroupName = "TLADe", Order = 3)]
        public bool SendIntraBars { get; set; } = false;

        [Display(Name = "Push daily bar at session rollover", GroupName = "TLADe", Order = 4)]
        public bool SendDailyBars { get; set; } = false;

        public TLAdeBridgeATAS() : base(true)
        {
            DataSeries[0].IsHidden = true;
            // Start the worker that sends bars sequentially
            Task.Run(BarWorker);
        }

        protected override void OnCalculate(int bar, decimal value)
        {
            // ATAS v8: bar = bar being processed; CurrentBar = total count; live bar = CurrentBar - 1
            var isLastBar = bar == CurrentBar - 1;

            // INIT: first time we reach the live bar
            if (!_liveMode && isLastBar)
            {
                var candle0 = GetCandle(bar);
                _liveMode   = true;
                Log($"INIT v2.6.0 bar={bar} CurrentBar={CurrentBar} price={candle0?.Close} instrument={DataProvider.InstrumentInfo.Instrument} port={Port}");
            }

            // 1) SPOT — throttled to SpotThrottleSeconds, sent from the live bar
            if (_liveMode && isLastBar)
            {
                var now    = DateTime.UtcNow;
                var candle = GetCandle(bar);
                var valueD = (double)value;
                var closeD = candle != null ? (double)candle.Close : 0.0;
                // value = tick real-time (actualizat per tick); candle.Close poate fi stale
                var price  = valueD > 0 ? valueD : closeD;

                if ((now - _lastSpotTime).TotalSeconds >= SpotThrottleSeconds)
                {
                    Log($"SPOT_DBG val={valueD} close={closeD} price={price} last={_lastPostedPrice}");
                    if (price > 0)
                    {
                        _lastPostedPrice = price;
                        _lastSpotTime    = now;
                        var ticker = NormalizeTicker(DataProvider.InstrumentInfo.Instrument);
                        Task.Run(() => PostSpot(ticker, price, now));
                    }
                }
            }

            // 2) BARA INCHISA — doar daca SendIntraBars=true (default false = NT8-mode)
            if (_liveMode && SendIntraBars && !isLastBar)
            {
                var candle = GetCandle(bar);
                if (candle == null) return;

                // TryAdd returns false if the bar has already been sent
                if (!_postedBarTimes.TryAdd(candle.Time.Ticks, 0)) return;

                // New-day detection -> send the D1 bar of the previous day
                if (SendDailyBars
                    && _lastBarDate != DateTime.MinValue.Date
                    && candle.Time.Date != _lastBarDate
                    && bar > 0
                    && (bar - 1) != _lastPostedDailyBar)
                {
                    _lastPostedDailyBar = bar - 1;
                    var prevCandle = GetCandle(bar - 1);
                    if (prevCandle != null)
                    {
                        var dailyCd = BuildCandleData(prevCandle);
                        var t       = NormalizeTicker(DataProvider.InstrumentInfo.Instrument);
                        Task.Run(() => PostDailyBar(t, bar - 1, dailyCd));
                    }
                }

                _lastBarDate = candle.Time.Date;

                var cd     = BuildCandleData(candle);
                var ticker = NormalizeTicker(DataProvider.InstrumentInfo.Instrument);
                _barChannel.Writer.TryWrite((ticker, bar, cd));
            }
        }

        // Single worker that consumes the channel sequentially — no thread-pool explosion
        private async Task BarWorker()
        {
            await foreach (var (ticker, bar, cd) in _barChannel.Reader.ReadAllAsync())
                await PostBar(ticker, bar, cd);
        }

        public override void Dispose()
        {
            _barChannel.Writer.Complete();
            base.Dispose();
        }

        private static string NormalizeTicker(string symbol)
        {
            var s = (symbol ?? "").ToUpperInvariant();
            if (s.Contains("NQ") || s.Contains("NDX") || s.Contains("MNQ"))
                return "NQ";
            return "ES";
        }

        private static CandleData BuildCandleData(IndicatorCandle candle)
        {
            return new CandleData
            {
                Open      = (double)candle.Open,
                High      = (double)candle.High,
                Low       = (double)candle.Low,
                Close     = (double)candle.Close,
                Volume    = (double)candle.Volume,
                AskVolume = (double)candle.Ask,
                BidVolume = (double)candle.Bid,
                Delta     = (double)candle.Delta,
                MaxDelta  = (double)candle.MaxDelta,
                MinDelta  = (double)candle.MinDelta,
                Time      = candle.Time
            };
        }

        private async Task PostSpot(string ticker, double price, DateTime ts)
        {
            var ci = System.Globalization.CultureInfo.InvariantCulture;
            var payload = "{" +
                $"\"ticker\":\"{ticker}\"," +
                $"\"spot\":{price.ToString(ci)}," +
                $"\"ts\":\"{ts:O}\"," +
                $"\"provider\":\"atas\"" +
                "}";
            await Post($"http://localhost:{Port}/push_spot", payload,
                       $"SPOT {ticker} {price.ToString(ci)}");
        }

        private async Task PostBar(string ticker, int bar, CandleData cd)
        {
            var ci = System.Globalization.CultureInfo.InvariantCulture;
            var payload = "{" +
                $"\"ticker\":\"{ticker}\"," +
                $"\"bar_index\":{bar}," +
                $"\"time\":\"{cd.Time:O}\"," +
                $"\"open\":{cd.Open.ToString(ci)}," +
                $"\"high\":{cd.High.ToString(ci)}," +
                $"\"low\":{cd.Low.ToString(ci)}," +
                $"\"close\":{cd.Close.ToString(ci)}," +
                $"\"volume\":{cd.Volume.ToString(ci)}," +
                $"\"ask_volume\":{cd.AskVolume.ToString(ci)}," +
                $"\"bid_volume\":{cd.BidVolume.ToString(ci)}," +
                $"\"delta\":{cd.Delta.ToString(ci)}," +
                $"\"max_delta\":{cd.MaxDelta.ToString(ci)}," +
                $"\"min_delta\":{cd.MinDelta.ToString(ci)}," +
                $"\"provider\":\"atas\"" +
                "}";
            await Post($"http://localhost:{Port}/push_bar", payload,
                       $"BAR {ticker} {cd.Time:HH:mm} C={cd.Close.ToString(ci)} D={cd.Delta:+0;-0}");
        }

        private async Task PostDailyBar(string ticker, int bar, CandleData cd)
        {
            var ci = System.Globalization.CultureInfo.InvariantCulture;
            var payload = "{" +
                $"\"ticker\":\"{ticker}\"," +
                $"\"bar_index\":{bar}," +
                $"\"time\":\"{cd.Time:yyyy-MM-dd}\"," +
                $"\"open\":{cd.Open.ToString(ci)}," +
                $"\"high\":{cd.High.ToString(ci)}," +
                $"\"low\":{cd.Low.ToString(ci)}," +
                $"\"close\":{cd.Close.ToString(ci)}," +
                $"\"volume\":{cd.Volume.ToString(ci)}," +
                $"\"ask_volume\":{cd.AskVolume.ToString(ci)}," +
                $"\"bid_volume\":{cd.BidVolume.ToString(ci)}," +
                $"\"delta\":{cd.Delta.ToString(ci)}," +
                $"\"provider\":\"atas\"" +
                "}";
            await Post($"http://localhost:{Port}/push_daily", payload,
                       $"DAILY {ticker} {cd.Time:yyyy-MM-dd} C={cd.Close.ToString(ci)}");
        }

        private async Task Post(string url, string json, string label)
        {
            try
            {
                var content  = new StringContent(json, Encoding.UTF8, "application/json");
                var response = await _http.PostAsync(url, content);
                Log($"{label} -> {(response.IsSuccessStatusCode ? "OK" : $"FAIL {(int)response.StatusCode}")}");
            }
            catch (TaskCanceledException)
            {
                Log($"TIMEOUT: {label}");
            }
            catch (Exception ex)
            {
                Log($"ERR ({label}): {ex.Message}");
            }
        }

        private static void Log(string msg)
        {
            try
            {
                File.AppendAllText(_logPath,
                    $"{DateTime.Now:yyyy-MM-dd HH:mm:ss} [TLAdeBridge] {msg}{Environment.NewLine}");
            }
            catch { }
        }

        private struct CandleData
        {
            public double   Open, High, Low, Close;
            public double   Volume, AskVolume, BidVolume;
            public double   Delta, MaxDelta, MinDelta;
            public DateTime Time;
        }
    }
}
